include/clad/Differentiator/DynamicGraph.h

```diff
@@ -0,0 +1,136 @@
+#ifndef CLAD_DIFFERENTIATOR_DYNAMICGRAPH_H
+#define CLAD_DIFFERENTIATOR_DYNAMICGRAPH_H
+
+#include <algorithm>
+#include <functional>
+#include <iostream>
+#include <queue>
+#include <set>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+namespace clad {
+template <typename T> class DynamicGraph {
+private:
+  /// Storing nodes in the graph. The index of the node in the vector is used as
+  /// a unique identifier for the node in the adjacency list.
+  std::vector<T> m_nodes;
+
+  /// Store the nodes in the graph as an unordered map from the node to a
+  /// boolean indicating whether the node is processed or not. The second
+  /// element in the pair is the id of the node in the nodes vector.
+  std::unordered_map<T, std::pair<bool, size_t>> m_nodeMap;
+
+  /// Store the adjacency list for the graph. The adjacency list is a map from
+  /// a node to the set of nodes that it has an edge to. We use integers inside
+  /// the set to avoid copying the nodes.
+  std::unordered_map<size_t, std::set<size_t>> m_adjList;
+
+  /// Set of source nodes in the graph.
+  std::set<size_t> m_sources;
+
+  /// Store the id of the node being processed right now.
+  int m_currentId = -1; // -1 means no node is being processed.
+
+  /// Maintain a queue of nodes to be processed next.
+  std::queue<size_t> m_toProcessQueue;
+
+public:
+  DynamicGraph() = default;
+
+  /// @brief  Add an edge from the source node to the destination node.
+  /// @param src
+  /// @param dest
+  void addEdge(const T& src, const T& dest) {
+    std::pair<bool, size_t> srcInfo = addNode(src);
+    std::pair<bool, size_t> destInfo = addNode(dest);
+    size_t srcId = srcInfo.second;
+    size_t destId = destInfo.second;
+    m_adjList[srcId].insert(destId);
+  }
+
+  /// @brief Add a node to the graph. If the node is already present, return the
+  /// id of the node in the graph. If the node is a source node, add it to the
+  /// queue of nodes to be processed.
+  /// @param node
```

Comment:
```suggestion
  /// \param node
```

That's common elsewhere in the codebase and llvm.